Lib/multiprocessing/queues.py

```diff
@@ -139,13 +139,10 @@ def put_nowait(self, obj):
 
     def close(self):
         self._closed = True
-        try:
-            self._reader.close()
-        finally:
-            close = self._close
-            if close:
-                self._close = None
-                close()
+        close = self._close
+        if close:
+            self._close = None
+            close()
```

Comment:
The change generally makes sense to me: we move closing the reader into the thread, so that if things get written after this code runs but before the thread gets around to cleaning itself up, the reader isn't closed prematurely.

But what I'm worried about is that we don't close the reader at all if there is no `self._close`. My reading is that that's safe because if there is a thread, then there's always also a `self._close`, and if there is no thread, no data will ever have been written to the pipe either. Is that right?

Reply:
> But what I'm worried about is that we don't close the reader at all if there is no self._close. My reading is that that's safe because if there is a thread, then there's always also a self._close, and if there is no thread, no data will ever have been written to the pipe either. Is that right?

That’s correct. `self._close` is initialized to `None` and the first `Queue.put` call sets `self._close` to a finalizer. So `Queue.close` will close the pipe ends *only* if some data have been written to the pipe. But `Queue.close` also sets `self._closed` to `True` and since both `Queue.get` and `Queue.put` raise a `ValueError` in that case, no data can be written to the pipe anyway after a `Queue.close` call. So it does not change the specification of [`Queue.close`](https://docs.python.org/3/library/multiprocessing.html#multiprocessing.Queue.close):

> Indicate that no more data will be put on this queue by the current process.